src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -392,45 +329,81 @@ EmberAfNodeOperationalCertStatus ConvertToNOCResponseStatus(CHIP_ERROR err)
 {
     emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Remove all Fabrics");
-    PlatformMgr().ScheduleWork(DoRemoveAllFabrics, 0);
     emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
     return true;
 }
 
 bool emberAfOperationalCredentialsClusterAddNOCCallback(EndpointId endpoint, app::CommandHandler * commandObj, ByteSpan NOCArray,
-                                                        ByteSpan IPKValue, NodeId CaseAdminNode, uint16_t AdminVendorId)
+                                                        ByteSpan IPKValue, NodeId adminNodeId, uint16_t adminVendorId)
 {
     EmberAfNodeOperationalCertStatus nocResponse = EMBER_ZCL_NODE_OPERATIONAL_CERT_STATUS_SUCCESS;
 
-    CHIP_ERROR err      = CHIP_NO_ERROR;
-    uint8_t fabricIndex = 0;
+    CHIP_ERROR err          = CHIP_NO_ERROR;
+    FabricIndex fabricIndex = 0;
 
     emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: commissioner has added an Op Cert");
 
+    err = gFabricBeingCommissioned.SetOperationalCertsFromCertArray(NOCArray);
+    VerifyOrExit(err == CHIP_NO_ERROR, nocResponse = ConvertToNOCResponseStatus(err));
+
+    gFabricBeingCommissioned.SetVendorId(adminVendorId);
+
+    err = GetGlobalFabricTable().AddNewFabric(gFabricBeingCommissioned, fabricIndex);
+    VerifyOrExit(err == CHIP_NO_ERROR, nocResponse = ConvertToNOCResponseStatus(err));
+
+    err = GetGlobalFabricTable().Store(fabricIndex);
+    VerifyOrExit(err == CHIP_NO_ERROR, nocResponse = ConvertToNOCResponseStatus(err));
+
+    // We have a new operational identity and should start advertising it.  We
+    // can't just wait until we get network configuration commands, because we
+    // might be on the operational network already, in which case we are
+    // expected to be live with our new identity at this point.
+    chip::app::Mdns::AdvertiseOperational();
+
+exit:
+
+    gFabricBeingCommissioned.Reset();
+    SendNOCResponse(commandObj, nocResponse, fabricIndex, ByteSpan());
+
+    if (nocResponse != EMBER_ZCL_NODE_OPERATIONAL_CERT_STATUS_SUCCESS)
+    {
+        emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Failed AddNOC request. Status %d", nocResponse);
+    }
+
+    return true;
+}
+
+bool emberAfOperationalCredentialsClusterUpdateNOCCallback(EndpointId endpoint, app::CommandHandler * commandObj, ByteSpan NOCArray)
+{
+    EmberAfNodeOperationalCertStatus nocResponse = EMBER_ZCL_NODE_OPERATIONAL_CERT_STATUS_SUCCESS;
+
+    CHIP_ERROR err          = CHIP_NO_ERROR;
+    FabricIndex fabricIndex = 0;
+
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: commissioner has updated the Op Cert");
+
+    // Fetch current fabric
     FabricInfo * fabric = retrieveCurrentFabric();
-    VerifyOrExit(fabric != nullptr, nocResponse = EMBER_ZCL_NODE_OPERATIONAL_CERT_STATUS_MISSING_CSR);
+    VerifyOrExit(fabric != nullptr, nocResponse = ConvertToNOCResponseStatus(CHIP_ERROR_INVALID_FABRIC_ID));
 
     err = fabric->SetOperationalCertsFromCertArray(NOCArray);
```

Comment:
Need a TODO to address the normative requirements to validate NOCArray. The certificate may not be valid.

Reply:
There's an outstanding issue for this: https://github.com/project-chip/connectedhomeip/issues/8433